src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -0,0 +1,262 @@
+            {
+                endpoints.Add((RouteEndpoint)CreateRouteEndpointBuilder(entry).Build());
+            }
+            return endpoints;
+        }
+    }
+
+    public override IReadOnlyList<RouteEndpoint> GetEndpointGroup(RouteGroupContext context)
+    {
+        var endpoints = new List<RouteEndpoint>(_routeEntries.Count);
+        foreach (var entry in _routeEntries)
+        {
+            endpoints.Add((RouteEndpoint)CreateRouteEndpointBuilder(entry, context.Prefix, context.Conventions).Build());
+        }
+        return endpoints;
+    }
+
+    public override IChangeToken GetChangeToken() => NullChangeToken.Singleton;
+
+    // For testing
+    internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+    {
+        if (_routeEntries.Count is not 1)
+        {
+            throw new InvalidOperationException($"There are {_routeEntries.Count} endpoints defined! This can only be called for a single endpoint.");
+        }
+
+        return CreateRouteEndpointBuilder(_routeEntries[0]);
+    }
+
+    [UnconditionalSuppressMessage("Trimmer", "IL2026",
+        Justification = "We surface a RequireUnreferencedCode in the call to Map method adding this EndpointDataSource. " +
+                        "The trimmer is unable to infer this.")]
+    private RouteEndpointBuilder CreateRouteEndpointBuilder(RouteEntry entry, RoutePattern? groupPrefix = null, IReadOnlyList<Action<EndpointBuilder>>? groupConventions = null)
+    {
+        var pattern = RoutePatternFactory.Combine(groupPrefix, entry.RoutePattern);
+        var handler = entry.RouteHandler;
+        var displayName = pattern.RawText ?? pattern.DebuggerToString();
+
+        // Methods defined in a top-level program are generated as statics so the delegate target will be null.
+        // Inline lambdas are compiler generated method so they be filtered that way.
+        if (GeneratedNameParser.TryParseLocalFunctionName(handler.Method.Name, out var endpointName)
+            || !TypeHelper.IsCompilerGeneratedMethod(handler.Method))
+        {
+            endpointName ??= handler.Method.Name;
+            displayName = $"{displayName} => {endpointName}";
+        }
+
+        if (entry.HttpMethods is not null)
+        {
+            // Prepends the HTTP method to the DisplayName produced with pattern + method name
+            displayName = $"HTTP: {string.Join(", ", entry.HttpMethods)} {displayName}";
+        }
+
+        if (entry.IsFallback)
+        {
+            displayName = $"Fallback {displayName}";
+        }
+
+        RequestDelegate? factoryCreatedRequestDelegate = null;
+        RequestDelegate redirectedRequestDelegate = context =>
+        {
+            if (factoryCreatedRequestDelegate is null)
+            {
+                throw new InvalidOperationException(Resources.RouteEndpointDataSource_RequestDelegateCannotBeCalledBeforeBuild);
+            }
+
+            return factoryCreatedRequestDelegate(context);
+        };
```

Comment:
What is this for? How would it show up?

Reply:
I see now. If a convention decorates the request delegate. Then we can't set overwrite the instance. The null check is just an extra safeguard but in practice nobody would call this before it's built.